File: ADT/djikstra's algo/2642.py
import heapq
from typing import List, Tuple, Dict

class Graph:
    def __init__(self, n: int, edges: List[List[int]]):
        self.N = n
        self.adj: Dict[int, List[Tuple[int, int]]] = {}
        for edge in edges:
            u, v, cost = edge
            if u not in self.adj:
                self.adj[u] = []
            self.adj[u].append((v, cost))
        self.pq: List[Tuple[int, int]] = []

    def addEdge(self, edge: List[int]):
        u, v, cost = edge
        if u not in self.adj:
            self.adj[u] = []
        self.adj[u].append((v, cost))

    def shortestPath(self, node1: int, node2: int) -> int:
        result = [float('inf')] * self.N
        result[node1] = 0
        self.pq = [(0, node1)]
        heapq.heapify(self.pq)

        while self.pq:
            d, node = heapq.heappop(self.pq)
            if d > result[node]:
                continue
            for adjNode, dist in self.adj.get(node, []):
                if d + dist < result[adjNode]:
                    result[adjNode] = d + dist
                    heapq.heappush(self.pq, (d + dist, adjNode))

        return -1 if result[node2] == float('inf') else result[node2]
        

File: ADT/djikstra's algo/2642.js
/**
 * @param {number} n
 * @param {number[][]} edges
 */
var Graph = function (n, edges) {
    this.n = n;
    this.adj = new Map();

    // initialize adjacency list
    for (let i = 0; i < n; i++) {
        this.adj.set(i, []);
    }

    // add initial edges
    for (const [u, v, w] of edges) {
        this.adj.get(u).push([v, w]);
    }
};

/** 
 * @param {number[]} edge
 * @return {void}
 */
Graph.prototype.addEdge = function (edge) {
    const [u, v, w] = edge;
    this.adj.get(u).push([v, w]);
};

/** 
 * @param {number} node1 
 * @param {number} node2
 * @return {number}
 */
Graph.prototype.shortestPath = function (node1, node2) {
    const dist = Array(this.n).fill(Infinity);
    dist[node1] = 0;

    // Min-heap: [distance, node]
    const pq = [[0, node1]];

    while (pq.length > 0) {
        // extract min
        pq.sort((a, b) => a[0] - b[0]);
        const [d, u] = pq.shift();

        if (u === node2) return d;
        if (d > dist[u]) continue;

        for (const [v, w] of this.adj.get(u)) {
            if (dist[v] > d + w) {
                dist[v] = d + w;
                pq.push([dist[v], v]);
            }
        }
    }

    return dist[node2] === Infinity ? -1 : dist[node2];
};
